src/passes/MultiMemoryLowering.cpp

```diff
@@ -0,0 +1,345 @@
+                                   : Literal(int64_t(offset))),
+        Builder::Mutable);
+      global->hasExplicitName = true;
+      wasm->addGlobal(std::move(global));
+    };
+
+    size_t offsetRunningTotal = 0;
+    // We don't need a page offset global for the first memory as it's always 0
+    // We also don't need a page offset global for the last memory as it's the
+    // combinedMemory we just created
+    for (Index i = 0; i < wasm->memories.size() - 1; i++) {
+      auto& memory = wasm->memories[i];
+      memoryIdxMap[memory->name] = i;
+      if (i != 0) {
+        Name name = Names::getValidGlobalName(
+          *wasm, std::string(memory->name.c_str()) + "_page_offset");
+        offsetGlobalNames.push_back(std::move(name));
+        addGlobal(name, offsetRunningTotal);
+      }
+      offsetRunningTotal += memory->initial;
+    }
+  }
+
+  void adjustActiveDataSegmentOffsets() {
+    Builder builder(*wasm);
+    ModuleUtils::iterActiveDataSegments(*wasm, [&](DataSegment* dataSegment) {
+      assert(dataSegment->offset->is<Const>());
+      auto iter = memoryIdxMap.find(dataSegment->memory);
+      assert(iter != memoryIdxMap.end());
+      dataSegment->memory = combinedMemory;
+      // No need to update the offset of data segments for the first memory
+      if (iter->second != 0) {
+        // We need to subtract 1 here because there is no offsetGlobal for the
+        // first memory
+        auto offsetGlobalName = offsetGlobalNames[iter->second - 1];
+        dataSegment->offset = builder.makeBinary(
+          pointerType == Type::i32 ? AddInt32 : AddInt64,
+          builder.makeGlobalGet(offsetGlobalName, pointerType),
+          dataSegment->offset);
+      }
+    });
+  }
+
+  void createMemorySizeFunctions() {
+    // Don't create a memory size function for the last memory in the vector as
+    // its the combinedMemory we just added
+    for (Index i = 0; i < wasm->memories.size() - 1; i++) {
+      auto function = memorySize(i);
+      memorySizeNames.push_back(function->name);
+      wasm->addFunction(std::move(function));
+    }
+  }
+
+  void createAdjustOffsetGlobalFunctions() {
+    // Don't create a memory size function for the last memory in the vector as
+    // its the combinedMemory we just added
+    for (Index i = 0; i < wasm->memories.size() - 1; i++) {
+      auto function = adjustOffsetGlobals(i);
+      adjustOffsetNames.push_back(function->name);
+      wasm->addFunction(std::move(function));
+    }
+  }
+
+  // This function replaces memory.grow instruction calls in the wasm module.
+  // Because the multiple discrete memories are lowered into a single memory,
+  // we need to adjust offsets as a particular memory receives an
+  // instruction to grow.
+  std::unique_ptr<Function> adjustOffsetGlobals(Index memIdx) {
+    Builder builder(*wasm);
+    Name functionName =
+      Names::getValidFunctionName(*wasm, "adjust_memory_offsets");
```

Comment:
It would be good to explicitly put the memory index into the function name. Right now the auto-generated indices appended to the names are off by one.